### PR TITLE
Implemented --query flag in `conan upload` (#2445)

### PR DESCRIPTION
**Pull request by joechrisellis**
You can now use the --query flag in the `conan upload` command like you
would in `conan remove`. In other words, you can use a query to specify
which packages will be uploaded to the server.

Will update documentation tonight.

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

### CODE REVIEW

File: conans/client/conan_api.py [hunk 1]
```diff
@@ -728,8 +728,9 @@ def search_packages(self, reference, query=None, remote_name=None, outdated=Fals
         return recorder.get_info()
 
     @api_method
-    def upload(self, pattern, package=None, remote_name=None, all_packages=False, force=False,
-               confirm=False, retry=2, retry_wait=5, skip_upload=False, integrity_check=False,
+    def upload(self, pattern, package=None, query=None, remote_name=None,
```
Comment (memsharded) [inline conans/client/conan_api.py]: I would say add the new argument in the last position, less likely to break to users of the API

**Comment by joechrisellis** [comment]
Would be happy to take a look at creating tests, will do this asap. :)

File: conans/test/command/upload_test.py [hunk 2]
```diff
@@ -123,13 +167,14 @@ def gzopen_patched(name, mode="r", fileobj=None, compresslevel=None, **kwargs):
             self.assertIn("ERROR: Error gzopen conan_package.tgz", client.out)
 
             export_folder = client.client_cache.package(package_ref)
+            print(export_folder)
```
Comment (lasote) [inline conans/test/command/upload_test.py]: Remove print

File: conans/client/command.py [hunk 3]
```diff
@@ -55,19 +55,22 @@ def __call__(self, parser, namespace, values, option_string=None):
...
 _BUILD_FOLDER_HELP = ("Directory for the build process. Defaulted to the current directory. A "
                       "relative path to current directory can also be specified")
 _INSTALL_FOLDER_HELP = ("Directory containing the conaninfo.txt and conanbuildinfo.txt files "
                         "(from previous 'conan install'). Defaulted to --build-folder")
 _KEEP_SOURCE_HELP = ("Do not remove the source folder in local cache, even if the recipe changed. "
                      "Use this for testing purposes only")
-_PATTERN_OR_REFERENCE_HELP = ("Pattern or package recipe reference, e.g., 'boost/*', "
-                              "'MyPackage/1.2@user/channel'")
+_PATTERN_OR_REFERENCE_HELP = ("Pattern or package recipe reference, e.g., '{}', "
+                              "'{}'".format(_REFERENCE_EXAMPLE, _PATTERN_EXAMPLE))
```
Comment (danimtb) [inline conans/client/command.py]: I'd rather use `%s` to format the string to be consistent with Conan codebase

File: conans/client/command.py [hunk 4]
```diff
@@ -55,19 +55,22 @@ def __call__(self, parser, namespace, values, option_string=None):
...
 _PATH_HELP = ("Path to a folder containing a conanfile.py or to a recipe file "
               "e.g., my_folder/conanfile.py")
-_QUERY_HELP = ("Packages query: 'os=Windows AND (arch=x86 OR compiler=gcc)'. The "
-               "'pattern_or_reference' parameter has to be a reference: MyPackage/1.2@user/channel")
+_QUERY_HELP = ("Packages query: '{}'. The 'pattern_or_reference' parameter has "
+               "to be a reference: {}".format(_QUERY_EXAMPLE, _REFERENCE_EXAMPLE))
```
Comment (danimtb) [inline conans/client/command.py]: Same thing here

**Comment by joechrisellis** [comment]
Removed extraneous print statement and removed new-style formatting for consistency. :)

**Comment by joechrisellis** [comment]
Made changes in conan-io/docs for the new option -- see: https://github.com/conan-io/docs/pull/772

**Comment by memsharded** [comment]
Merged! Will be in conan 1.7. Thanks again for your contribution :)

**Comment by joechrisellis** [comment]
Awesome, and no problem :D

**Pull request merged**